### ISSUE TITLE
reposync doesn't work in Fedora 32

### ISSUE THREAD

**Issue by tpw56j** [cobbler/cobbler#2378]
### Describe the bug
Several problems in the code reposync.py in Fedora 32

### To Reproduce

1. # cobbler reposync
2. See errors

### Expected behavior
Synchronize packages of a remote DNF repository to a local directory.

### Version

3.1.2

### Additional context

**Problem 1:**
dnf reposync (as in dnf in fedora or el8) uses --download-path

running: /usr/bin/dnf reposync --newest-only --delete --config=/var/www/cobbler/repo_mirror/OEL-7-Latest/.origin/OEL-7-Latest.repo --repoid=OEL-7-Latest --download_path=/var/www/cobbler/repo_mirror -a x86_64
received on stdout: 
received on stderr: usage: dnf reposync [-c [config file]] [-q] [-v] [--version]
                    [--installroot [path]] [--nodocs] [--noplugins]
                    [--enableplugin [plugin]] [--disableplugin [plugin]]
                    [--releasever RELEASEVER] [--setopt SETOPTS]
--nobest] [-C]
                    [-R [minutes]] [-d [debug level]] [--debugsolver]
                    [--showduplicates] [-e ERRORLEVEL] [--obsoletes]
                    [--rpmverbosity [debug level name]] [-y] [--assumeno]
--repo
--disable] [-x [package]]
                    [--disableexcludes [repo]] [--repofrompath [repo,path]]
                    [--noautoremove] [--nogpgcheck] [--color COLOR]
                    [--refresh] [-4] [-6] [--destdir DESTDIR] [--downloadonly]
                    [--comment COMMENT] [--bugfix] [--enhancement]
                    [--newpackage] [--security] [--advisory ADVISORY]
                    [--bz BUGZILLA] [--cve CVES]
                    [--sec-severity {Critical,Important,Moderate,Low}]
                    [--forcearch ARCH] [-a [arch]] [--delete] [-m]
                    [--download-metadata] [-n] [-p DOWNLOAD_PATH]
                    [--norepopath] [--metadata-path METADATA_PATH] [--source]
                    [--remote-time] [-u]
dnf reposync: error: unrecognized arguments: --download_path=/var/www/cobbler/repo_mirror

**Problem 2:**
Uninitialized variables cert, verify

received on stderr: 
Exception occured: <class 'UnboundLocalError'>
Exception value: local variable 'cert' referenced before assignment
Exception Info:
  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 163, in run
    self.sync(repo)

  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 206, in sync
    self.yum_sync(repo)

  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 524, in yum_sync
    (cert, verify) = self.gen_urlgrab_ssl_opts(repo.yumopts)

  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 434, in gen_urlgrab_ssl_opts
    return (cert, verify)

Workaround:
Assign a value to variables prior to use.
--- reposync.py.orig	2020-07-06 20:58:12.496673490 +0300
+++ reposync.py.my	2020-07-17 09:52:50.525167207 +0300
@@ -421,6 +421,8 @@
         :rtype: (str, bool)
         """
         # use SSL options if specified in yum opts
+        cert = None
+        verify = False
         if 'sslclientkey' and 'sslclientcert' in yumopts:
             cert = (yumopts['sslclientcert'], yumopts['sslclientkey'])
         # Note that the default of requests is to verify the peer and host but the default here is NOT to verify them

**Problem 3:**
The download_file definition does not contain the "verify" parameter.
download_manager.py:    def download_file(self, url, dst, proxies=None, cert=None)

received on stderr: 
Exception occured: <class 'TypeError'>
Exception value: can only concatenate str (not "tuple") to str
Exception Info:
  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 163, in run
    self.sync(repo)

  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 206, in sync
    self.yum_sync(repo)

  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 530, in yum_sync
    utils.die(self.logger, "failed to fetch " + src + " " + e.args)

Workaround:

--- reposync.py.orig	2020-07-06 20:58:12.496673490 +0300
+++ reposync.py	2020-07-17 10:22:14.495238163 +0300
@@ -523,7 +525,7 @@
         dst = temp_path + "/repomd.xml"
         (cert, verify) = self.gen_urlgrab_ssl_opts(repo.yumopts)
         try:
-            self.dlmgr.download_file(src, dst, proxies, cert, verify)
+            self.dlmgr.download_file(src, dst, proxies, cert)
         except Exception as e:
             utils.die(self.logger, "failed to fetch " + src + " " + e.args)

**Problem 4:**
YUM API has been removed in Fedora 32.

# dnf install yum-utils
Package dnf-utils-4.0.16-1.fc32.noarch is already installed.
Dependencies resolved.
Nothing to do.
Complete!

# cobbler reposync
< .. >
received on stderr: 
Exception occured: <class 'NameError'>
Exception value: name 'yum' is not defined
Exception Info:
  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 163, in run
    self.sync(repo)

  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 206, in sync
    self.yum_sync(repo)

  File "/usr/lib/python3.8/site-packages/cobbler/actions/reposync.py", line 535, in yum_sync
    rmd = yum.repoMDObject.RepoMD('', "%s/repomd.xml" % (temp_path))

**Comment by SchoolGuy** [comment]
Thanks for spotting and debugging this. If you could submit a PR with the most problematic fixes I would be very grateful. For the replacement of yum to dnf I don't have any timeline as this is something bigger and for that we have a lot of technical debt... Also we are planning to use pulp in the future to replace at least some of the code in `reposync.py`.

**Comment by tpw56j** [comment]
I will try to submit a PR.
As far as I understand, for this you need to first create a branch?

**Comment by SchoolGuy** [comment]
@tpw56j Yes but before that you need to fork the project. Although it is a bit to read I would recommend you to read this guide: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/proposing-changes-to-your-work-with-pull-requests

If you need more help don't worry I have more documents and videos on my bookmark list.

**Comment by SchoolGuy** [comment]
The TODO list to close this issue is the following:

- [ ] Adjust specfile to include librepo
- [ ] Adjust docs for installation
- [ ] Remove old yum code from `reposync.py`
- [ ] Update other documentation on rtd to give more details about our repository features.
